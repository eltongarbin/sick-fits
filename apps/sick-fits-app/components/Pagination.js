import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { useQuery } from '@apollo/client';

import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    productsCount
  }
`;

export default function Pagination({ page }) {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);

  if (loading) return 'Loading...';
  if (error) return <DisplayError error={error} />;

  const count = data.productsCount;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles data-testid="pagination">
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`} aria-disabled={page <= 1}>
        ← Prev
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`} aria-disabled={page >= pageCount}>
        Next →
      </Link>
    </PaginationStyles>
  );
}
