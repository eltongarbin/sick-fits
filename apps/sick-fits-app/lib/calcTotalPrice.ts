export const calcTotalPrice = (cart: any) =>
  cart.reduce((tally: number, cartItem: any) => {
    if (!cartItem.product) return tally;
    return tally + cartItem.quantity * cartItem.product.price;
  }, 0);
