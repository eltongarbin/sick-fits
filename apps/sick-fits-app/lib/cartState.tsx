/* eslint-disable @typescript-eslint/no-empty-function */
import { createContext, ReactNode, useContext, useState } from 'react';

const LocalStateContext = createContext({
  cartOpen: false,
  toggleCart: () => {},
  closeCart: () => {},
  openCart: () => {},
});

const { Provider: LocalStateProvider } = LocalStateContext;

type CartStateProviderProps = {
  children: ReactNode;
};

export function CartStateProvider({ children }: CartStateProviderProps) {
  const [cartOpen, setCartOpen] = useState(false);

  const toggleCart = () => setCartOpen(!cartOpen);
  const closeCart = () => setCartOpen(false);
  const openCart = () => setCartOpen(true);

  return (
    <LocalStateProvider value={{ cartOpen, toggleCart, closeCart, openCart }}>
      {children}
    </LocalStateProvider>
  );
}

export function useCart() {
  return useContext(LocalStateContext);
}
