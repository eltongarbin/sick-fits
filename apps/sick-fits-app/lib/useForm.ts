import { ChangeEvent, useState } from 'react';

export const useForm = (initial = {}) => {
  const [inputs, setInputs] = useState(initial);
  // const initialValues = useMemo(
  //   () => Object.values(initial).join(''),
  //   [initial]
  // );

  // useEffect(() => {
  //   setInputs(initial);
  // }, [initialValues]);

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    const { value, name, type } = e.target;
    let newValue;

    if (type === 'number') {
      newValue = parseInt(value);
    } else if (type === 'file') {
      newValue = e.target.files?.item(0);
    } else {
      newValue = value;
    }

    setInputs({
      ...inputs,
      [name]: newValue,
    });
  };

  const resetForm = () => {
    setInputs(initial);
  };

  const clearForm = () => {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );

    setInputs(blankState);
  };

  return { inputs, handleChange, resetForm, clearForm };
};
