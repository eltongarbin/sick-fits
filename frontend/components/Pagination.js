import React from 'react';
import PropTypes from 'prop-types';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/react-hooks';
import Head from 'next/head';
import Link from 'next/link';

import { perPage } from '../config';
import PaginationStyles from './styles/PaginationStyles';
import ErrorMessage from './ErrorMessage';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = ({ page }) => {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);

  if (error) return <ErrorMessage error={error} />;
  if (loading) return <p>Loading....</p>;

  const { count } = data.itemsConnection.aggregate;
  const pages = Math.ceil(count / perPage);

  return (
    <PaginationStyles>
      <Head>
        <title>
          Sick Fits! - Page {page} of {pages}
        </title>
      </Head>
      <Link
        prefetch
        href={{
          pathname: 'items',
          query: {
            page: page - 1
          }
        }}
      >
        <a className="prev" aria-disabled={page <= 1}>
          ◀️ Prev
        </a>
      </Link>
      <p>
        Page {page} of {pages}!
      </p>
      <p>{count} Items Total</p>
      <Link
        prefetch
        href={{
          pathname: 'items',
          query: {
            page: page + 1
          }
        }}
      >
        <a className="prev" aria-disabled={page >= pages}>
          Next ▶️
        </a>
      </Link>
    </PaginationStyles>
  );
};

Pagination.propTypes = {
  page: PropTypes.number.isRequired
};

export default Pagination;
