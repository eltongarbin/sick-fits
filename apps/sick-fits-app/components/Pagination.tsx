import gql from 'graphql-tag';
import Head from 'next/head';
import Link from 'next/link';
import { useQuery } from '@apollo/client';

import { PaginationStyles } from './Pagination.styles';
import { ErrorMessage } from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    productsCount
  }
`;

type PaginationProps = {
  page: number;
};

export const Pagination = ({ page }: PaginationProps) => {
  const { data, loading, error } = useQuery(PAGINATION_QUERY);

  if (loading) return <>Loading...</>;
  if (error) return <ErrorMessage error={error} />;

  const count = data.productsCount;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationStyles data-testid="pagination">
      <Head>
        <title>
          Sick Fits - Page {page} of {pageCount}
        </title>
      </Head>
      <Link href={`/products/${page - 1}`} aria-disabled={page <= 1}>
        ← Prev
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>{count} Items Total</p>
      <Link href={`/products/${page + 1}`} aria-disabled={page >= pageCount}>
        Next →
      </Link>
    </PaginationStyles>
  );
};
