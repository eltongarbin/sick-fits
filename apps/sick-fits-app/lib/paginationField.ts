import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // tells apollo we will take care of everything
    read(existing = [], { args, cache }) {
      const { skip, first } = args;

      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?.productsCount?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing.slice(skip, skip + first).filter((x) => x);

      if (items.length && items.length !== first && page === pages)
        return items;

      if (items.length !== first) return false;

      if (items.length) return items;

      return false;
    },
    merge(existing, incoming, { args }) {
      // This runs when the Apollo client comes back from the network with our product
      const { skip } = args;
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i)
        merged[i] = incoming[i - skip];

      return merged;
    },
  };
}
