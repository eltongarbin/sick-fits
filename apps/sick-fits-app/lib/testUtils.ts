import casual from 'casual';
import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  // __typename: 'Item',
  id: 'abc123',
  price: 5000,
  user: null,
  photo: {
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
    altText: 'dogs are best',
  },
  name: 'dogs are best',
  description: 'dogs',
});

const fakeUser = (overrides?: Record<string, unknown>) => ({
  __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = (overrides?: Record<string, unknown>) => ({
  __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  store: Record<string, string> = {};

  clear() {
    this.store = {};
  }

  getItem(key: string) {
    return this.store[key] || null;
  }

  setItem(key: string, value: string) {
    this.store[key] = value.toString();
  }

  removeItem(key: string) {
    delete this.store[key];
  }
}

const makePaginationMocksFor = (length: number) => [
  {
    request: { query: PAGINATION_QUERY },
    result: {
      data: {
        productsCount: length,
      },
    },
  },
];

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
