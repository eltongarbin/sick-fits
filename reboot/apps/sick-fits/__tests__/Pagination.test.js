import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";

import Pagination from "../components/Pagination";
import { makePaginationMocksFor } from "../lib/testUtils";

describe("<Pagination />", () => {
  it("displays a loading message", () => {
    render(
      <MockedProvider mocks={makePaginationMocksFor(1)}>
        <Pagination />
      </MockedProvider>
    );
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  it("renders pagination for 18 items", async () => {
    const { container } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    await screen.findByTestId("pagination");
    expect(screen.getByText("Page 1 of 5")).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("disables the prev page on first page", async () => {
    render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    await screen.findByTestId("pagination");

    expect(screen.getByText(/Prev/)).toHaveAttribute("aria-disabled", "true");
    expect(screen.getByText(/Next/)).toHaveAttribute("aria-disabled", "false");
  });

  it("disables the next page on last page", async () => {
    render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );

    await screen.findByTestId("pagination");

    expect(screen.getByText(/Prev/)).toHaveAttribute("aria-disabled", "false");
    expect(screen.getByText(/Next/)).toHaveAttribute("aria-disabled", "true");
  });

  it("enables all on middle page", async () => {
    render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={2} />
      </MockedProvider>
    );

    await screen.findByTestId("pagination");

    expect(screen.getByText(/Prev/)).toHaveAttribute("aria-disabled", "false");
    expect(screen.getByText(/Next/)).toHaveAttribute("aria-disabled", "false");
  });
});
