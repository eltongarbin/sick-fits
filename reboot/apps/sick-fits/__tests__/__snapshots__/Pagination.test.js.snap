// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> renders pagination for 18 items 1`] = `
<div
  class="sc-bcXHqe frpvnm"
  data-testid="pagination"
>
  <a
    aria-disabled="true"
    href="/products/0"
  >
    ← Prev
  </a>
  <p>
    Page 
    1
     of 
    5
  </p>
  <p>
    18
     Items Total
  </p>
  <a
    aria-disabled="false"
    href="/products/2"
  >
    Next →
  </a>
</div>
`;
